### ISSUE TITLE
payu profiler status

### ISSUE THREAD

**Issue by manodeep** [payu-org/payu#544]
I see that there is a `profiler` sub-module within payu - the code however looks sparse and I could not find any relevant documentation about the profiler - neither for further development nor for how to use it. The only relevant issue I found is [this one](https://github.com/payu-org/payu/issues/27) - which seems to imply that the profiler code might be unused/dead code.

**Comment by aidanheerdegen** [comment]
The profiler code was mostly added by @marshallward when he needed it. It isn't actively used or tested as far as I know.

That said the machinery is there to do things like preload libraries and wrap `mpirun` commands.

Can you give some details of what you require and we can see if it fits in the current design?

**Comment by manodeep** [comment]
Thanks @aidanheerdegen. 

I am hoping to use the Intel Vtune profiler - https://www.intel.com/content/www/us/en/docs/vtune-profiler/user-guide/2025-0/mpi-code-analysis.html

The summary is that something like this needs to run for the codes to be profiled:
`mpirun [mpirun-options] vtune [vtune-options] <program> [<args>]` (with possibly a `--` demarcating vtune options and the exe)

For example,
`mpirun -n 16 –ppn 4 –l vtune -collect hotspots -k sampling-mode=hw -trace-mpi -result-dir my_result -- my_app.exe`

**Comment by aidanheerdegen** [comment]
What happens with multiple executables, e.g.

```bash
mpirun --mca io ompio --mca io_ompio_num_aggregators 1 -wdir work/1deg_jra55_ryf-expt-acfa7666/atmosphere -np 1  work/1deg_jra55_ryf-expt-acfa7666/atmosphere/yatm.exe : -wdir work/1deg_jra55_ryf-expt-acfa7666/ocean -np 216  work/1deg_jra55_ryf-expt-acfa7666/ocean/fms_ACCESS-OM.x : -wdir work/1deg_jra55_ryf-expt-acfa7666/ice -np 24  work/1deg_jra55_ryf-expt-acfa7666/ice/cice_auscom_360x300_24x1_24p.exe
```

Does `vtune -collect hotspots -k sampling-mode=hw -trace-mpi -result-dir my_result --` have to be inserted before every executable (after the `-np` options)?

**Comment by manodeep** [comment]
I tried earlier today adding the vtune specific bit in front of every exe (and a new results directory) - but that failed with a seemingly unrelated error about too many files open (but the suggested fix didn't work either). Will try again tomorrow.

**Comment by manodeep** [comment]
Or possibly just use a single exe setup - which would be a lot more convenient.

**Comment by aidanheerdegen** [comment]
If we're going to add profiler support though, it would be good to know how it should be invoked.

**Comment by manodeep** [comment]
Of course! The docs show the single-executable setup; I will have to experiment to figure out the multi-executable one.

**Comment by marshallward** [comment]
I think it makes sense to gut all of the profile config.  Whatever might still be floating around wasn't going in the right direction.

If I were to redo it, I'd probably focus on dedicated wrapper support, with a mind towards profiler, debuggers, mutli-executable programs like ACCESS, etc.

**Comment by manodeep** [comment]
In a related attempt where I am trying to use valgrind, I got the too many files open warning open as well. However, that error disappeared when I first ran the `python payu-run` command. The steps were: 
1. `payu run <-f>` and then copy-paste the qsub command with the `-I` flag but removing the `-- <python> <payu-run>` bit
2. Once the interactive job starts, run the previously deleted `<python> <payu-run>`, and Ctrl-C when the `mpirun` command is invoked
3. Alter the mpirun command to suit the individual need
4. Make sure to exit the shell and terminate the interactive job

**Comment by aidanheerdegen** [comment]
> I think it makes sense to gut all of the profile config. Whatever might still be floating around wasn't going in the right direction.

It does seem like its presence is confusing, without any useful functionality. Have added a clean-up tag.

**Comment by aidanheerdegen** [comment]
> I think it makes sense to gut all of the profile config

Are you suggesting we get rid of the `profile` code completely, or just the bits that configure different profilers?

**Comment by marshallward** [comment]
> Are you suggesting we get rid of the `profile` code completely, or just the bits that configure different profilers?

Yes, I think it can all be removed.  It makes more sense to strengthen the application launcher to support profiler-like applications.

**Comment by KAUR1984** [comment]
Closing this issue as completed. Related PR - #783.
